File: app/src/main/java/com/synapse/social/studioasinc/PresenceManager.java
package com.synapse.social.studioasinc;

import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

/**
 * A utility class to manage the user's online presence in Firebase.
 * This class uses static methods and is not meant to be instantiated.
 */
public final class PresenceManager {

    private static final DatabaseReference database = FirebaseDatabase.getInstance().getReference("/status");

    // Private constructor to prevent instantiation.
    private PresenceManager() {}

    /**
     * Sets user status to "online" and registers onDisconnect to set to "offline".
     * Call when the app is brought to the foreground.
     * @param uid The Firebase UID of the current user.
     */
    public static void goOnline(String uid) {
        if (uid == null || uid.isEmpty()) return;
        DatabaseReference userStatusRef = database.child(uid);
        userStatusRef.setValue("online");
        // Set onDisconnect to mark the user as offline.
        userStatusRef.onDisconnect().setValue("offline");
    }

    /**
     * Explicitly sets the user's status to "offline".
     * Call on manual logout or when the app is backgrounded.
     * @param uid The Firebase UID of the current user.
     */
    public static void goOffline(String uid) {
        if (uid == null || uid.isEmpty()) return;
        database.child(uid).setValue("offline");
    }

    /**
     * Sets status to "chatting_with_<otherUserUid>".
     * Call from ChatActivity's onResume or onStart.
     * @param currentUserUid The UID of the current user.
     * @param otherUserUid The UID of the user they are chatting with.
     */
    public static void setChattingWith(String currentUserUid, String otherUserUid) {
        if (currentUserUid == null || currentUserUid.isEmpty()) return;
        database.child(currentUserUid).setValue("chatting_with_" + otherUserUid);
    }

    /**
     * Reverts the user's status back to "online".
     * Call from ChatActivity's onPause or onStop.
     * @param currentUserUid The UID of the current user.
     */
    public static void stopChatting(String currentUserUid) {
        if (currentUserUid == null || currentUserUid.isEmpty()) return;
        // Revert to "online" if the user is still connected.
        // The onDisconnect handler will manage "offline" on app close.
        database.child(currentUserUid).setValue("online");
    }
}

File: app/src/main/java/com/synapse/social/studioasinc/PresenceManager.kt
package com.synapse.social.studioasinc

import com.google.firebase.database.FirebaseDatabase

/**
 * Manages user online presence in Firebase, writing to the correct database path.
 * Handles online, offline (timestamp), and chat statuses.
 */
object PresenceManager {

    // Correct database reference to the 'users' node
    private val usersRef = FirebaseDatabase.getInstance().getReference("skyline/users")

    /**
     * Returns the specific database reference for a user's status.
     * Path: /skyline/users/{uid}/status
     */
    private fun getUserStatusRef(uid: String) = usersRef.child(uid).child("status")

    /**
     * Sets user status to "online".
     * Registers onDisconnect to set a timestamp for last seen.
     * @param uid The Firebase UID of the current user.
     */
    @JvmStatic
    fun goOnline(uid: String) {
        val statusRef = getUserStatusRef(uid)
        val activityRef = usersRef.child(uid).child("activity")
        statusRef.setValue("online")
        // On disconnect, set the last seen time as a timestamp string
        statusRef.onDisconnect().setValue(System.currentTimeMillis().toString())
        activityRef.onDisconnect().removeValue()
    }

    /**
     * Explicitly sets the user's status to a timestamp (last seen).
     * @param uid The Firebase UID of the current user.
     */
    @JvmStatic
    fun goOffline(uid: String) {
        // Set the last seen time as a timestamp string
        getUserStatusRef(uid).setValue(System.currentTimeMillis().toString())
    }

    /**
     * Sets status to "chatting_with_<otherUserUid>".
     * @param currentUserUid The UID of the current user.
     * @param otherUserUid The UID of the user they are chatting with.
     */
    @JvmStatic
    fun setChattingWith(currentUserUid: String, otherUserUid: String) {
        UserActivity.setActivity(currentUserUid, "chatting_with_$otherUserUid")
    }

    /**
     * Reverts the user's status back to "online".
     * @param currentUserUid The UID of the current user.
     */
    @JvmStatic
    fun stopChatting(currentUserUid: String) {
        UserActivity.clearActivity(currentUserUid)
    }

    @JvmStatic
    fun setActivity(uid: String, activity: String) {
        UserActivity.setActivity(uid, activity)
    }
}
